community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/TemporalIndexCache.java

```diff
@@ -0,0 +1,248 @@
+/*
+ * Copyright (c) 2002-2018 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.impl.index.schema;
+
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+import java.util.function.Function;
+
+import org.neo4j.helpers.collection.Iterators;
+import org.neo4j.values.storable.ValueGroup;
+
+/**
+ * Cache for lazily creating parts of the temporal index. Each part is created using the factory
+ * the first time it is selected in a select() query, or the first time it's explicitly
+ * asked for using e.g. date().
+ *
+ * Iterating over the cache will return all currently created parts.
+ *
+ * @param <T> Type of parts
+ * @param <E> Type of exception potentially thrown during creation
+ */
+class TemporalIndexCache<T, E extends Exception> implements Iterable<T>
+{
+    private final Factory<T, E> factory;
+
+    private volatile T date;
```

Comment:
Implementation may look cleaner and would allow for less duplication with an internal T[] instead?

Reply:
We tried this but it felt like roughly equally long, but slightly more complex. Decided to keep as is until we find some other reason to change it.